File: php/src/Node.php
<?php

namespace Acme;

class Node
{
    public int $value;
    public Node|null $left;
    public Node|null $right;

    public function __construct(int $value, Node|null $left, Node|null $right)
    {
        $this->value = $value;
        $this->left = $left;
        $this->right = $right;
    }

    public function num_nodes(): int
    {
        $count = 1;
        if ($this->left) {
            $count += $this->left->num_nodes();
        }
        if ($this->right) {
            $count += $this->right->num_nodes();
        }
        return $count;
    }

    public static function from_object(object $data): self
    {
        $left = $data->left ? static::from_object($data->left) : null;
        $right = $data->right ? static::from_object($data->right) : null;

        return new static($data->value, $left, $right);
    }
}

function binary_trees_equal(Node $tree1, Node $tree2): bool
{
    if ($tree1->value !== $tree2->value) {
        return false;
    }

    if (
        !isset($tree1->left) &&
        !isset($tree1->right) &&
        !isset($tree2->left) &&
        !isset($tree2->right)
    ) {
        return true;
    }

    if ((isset($tree1->left) && !isset($tree2->left)) ||
        (!isset($tree1->left) && isset($tree2->left)) ||
        (isset($tree1->right) && !isset($tree2->right)) ||
        (!isset($tree1->right) && isset($tree2->right))
    ) {
        return false;
    }

    $left_equal = true;
    $right_equal = true;

    if (isset($tree1->left) && isset($tree2->left)) {
        $left_equal = binary_trees_equal($tree1->left, $tree2->left);
    }

    if (isset($tree1->right) && isset($tree2->right)) {
        $right_equal = binary_trees_equal($tree1->right, $tree2->right);
    }

    return $left_equal && $right_equal;
}

File: rust/src/node.rs
use serde::Deserialize;

#[derive(PartialEq, Deserialize)]
pub struct Node {
    pub left: Option<Box<Node>>,
    pub right: Option<Box<Node>>,
    pub value: u64,
}

impl Node {
    pub fn num_nodes(&self) -> u64 {
        let mut count = 1;

        if let Some(left) = &self.left {
            count = count + left.num_nodes();
        }

        if let Some(right) = &self.right {
            count = count + right.num_nodes();
        }

        count
    }
}

pub fn binary_trees_equal(tree1: Node, tree2: Node) -> bool {
    if tree1.value != tree2.value {
        return false;
    }

    match (tree1.left, tree2.left, tree1.right, tree2.right) {
        (None, None, None, None) => true,
        (Some(_), None, None, None) => false,
        (None, Some(_), None, None) => false,
        (None, None, Some(_), None) => false,
        (None, None, None, Some(_)) => false,
        (Some(_), None, Some(_), None) => false,
        (Some(_), None, None, Some(_)) => false,
        (None, Some(_), Some(_), None) => false,
        (None, Some(_), None, Some(_)) => false,
        (Some(_), Some(_), Some(_), None) => false,
        (Some(_), Some(_), None, Some(_)) => false,
        (Some(_), None, Some(_), Some(_)) => false,
        (None, Some(_), Some(_), Some(_)) => false,
        (Some(one_left), Some(two_left), None, None) => binary_trees_equal(*one_left, *two_left),
        (None, None, Some(one_right), Some(two_right)) => {
            binary_trees_equal(*one_right, *two_right)
        }
        (Some(one_left), Some(two_left), Some(one_right), Some(two_right)) => {
            if binary_trees_equal(*one_left, *two_left) {
                return binary_trees_equal(*one_right, *two_right);
            }

            false
        }
    }
}
